# Data Manipulation with Pandas  
1. Transforming DataFrames
2. Aggregating DataFrames
3. Slicing and Indexing DataFrames
4. Creating and Visualizing DataFrames

### 1. Transforming DataFrames  
  
**What´s the point of pandas?**  
* Data Manipulation
 * Ch 1, Sorting and Subsetting, Creating new columns
 * Ch 2, Summary statistics, Counting, Grouped Summary statistics
 * Ch 3, Subsetting using slicing, Indexing and subsetting using index
* Data Visualization  
 * Ch 4, Plotting, Handling missing data, Reading data into a DataFrame
  
  
* **Introducing DataFrames**  
pandas is built on top of NumPy and Matplotlib.  
  
Rectangular data or tabular data is the a common form of data pandas is design to work with it.  
Each row is an observation and each column is a property or attribute. In pandas this rectangular data is represented as a `DataFrame` object. Other programming languages has something similar, `SQL` has `data table` and `R` also has `data.frame`. Every column have the same data type on each row but different columns can have different data types.  
  
Exploring a DataFrame object using methods/attributes:  
  
`df.head()` method displays the first rows (default is 5)  
`df.info()` method displays the names of columns, data types they contain and if there are any missing alues.  
`df.shape` attribute shows the number of rows followed by the number of clolumns, e.g `(8,7)`  
`df.describe()` method calculates some summary statistics for numerical columns.  
  
A `DataFrame` consists of three different components accessible using attributes.  
  
`df.values` attribute contains the data values in a 2D numpy array  
`df.columns` attribute contains the labels of the column names as index object.  
`df.index` attribute contains row numbers of row names as index object. Note: Row labes are stored in `.index` and not in `.rows`!  
  
**pandas Philosophy**  
There should be one -- and preferably only one -- obvious way to do it!  
*- The Zen of Python* by Tim Peters, Item 13 https://www.python.org/dev/peps/pep-0020/ 
Pandas deliberately does not follow this philosphoy, instead there are often many ways to solve a problem in pandas and we can choose the best one. pandas gives many tools to give options.

**Inspecting a DataFrame**  
When we have a new DataFrame to work with, the first thing you need to do is explore it and see what it contains. There are several useful methods and attributes for this.  
  
`.head()` returns the first few rows (the “head” of the `DataFrame`).  
`.info()` shows information on each of the columns, such as the data type and number of missing values.  
`.shape` returns the number of rows and columns of the DataFrame.  
`.describe()` calculates a few summary statistics for each column.

In [1]:
import pandas as pd

# OMA Need to create the DF
path = "data/homelessness.csv"
homelessness = pd.read_csv(path, index_col=0)

In [2]:
# Print the head of the homelessness data
print(homelessness.head())

               region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588


In [3]:
# Print information about homelessness
print(homelessness.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      51 non-null     int64  
 1   region          51 non-null     object 
 2   state           51 non-null     object 
 3   individuals     51 non-null     float64
 4   family_members  51 non-null     float64
 5   state_pop       51 non-null     int64  
dtypes: float64(2), int64(2), object(2)
memory usage: 2.5+ KB
None


In [4]:
# Print the shape of homelessness
print(homelessness.shape)

(51, 6)


In [6]:
# Print a description of homelessness
print(homelessness.describe())

       Unnamed: 0    individuals  family_members     state_pop
count   51.000000      51.000000       51.000000  5.100000e+01
mean    25.000000    7225.784314     3504.882353  6.405637e+06
std     14.866069   15991.025083     7805.411811  7.327258e+06
min      0.000000     434.000000       75.000000  5.776010e+05
25%     12.500000    1446.500000      592.000000  1.777414e+06
50%     25.000000    3082.000000     1482.000000  4.461153e+06
75%     37.500000    6781.500000     3196.000000  7.340946e+06
max     50.000000  109008.000000    52070.000000  3.946159e+07


We see the average number of homeless individuals of each state is about 7226.  
Now we do some more exploration of the df!  
  
**Parts of a DataFrame**  
To better understand DataFrame objects, it's useful to know that they consist of three components, stored as attributes:  
  
* `.values`: A two-dimensional NumPy array of values.  
* `.columns`: An index of columns: the column names.  
* `.index`: An index for the rows: either row numbers or row names.  
  
You can usually think of indexes as a list of strings or numbers, though the pandas `Index` data type allows for more sophisticated options. (These will be covered later in the course.)  
  
* Print a 2D NumPy array of the values in `homelessness`
* Print the column names of `homelessness`
* Print the index of `homelessness`

In [13]:
# Print the values of homelessness
print(homelessness.values)

[['East South Central' 'Alabama' 2570.0 864.0 4887681]
 ['Pacific' 'Alaska' 1434.0 582.0 735139]
 ['Mountain' 'Arizona' 7259.0 2606.0 7158024]
 ['West South Central' 'Arkansas' 2280.0 432.0 3009733]
 ['Pacific' 'California' 109008.0 20964.0 39461588]
 ['Mountain' 'Colorado' 7607.0 3250.0 5691287]
 ['New England' 'Connecticut' 2280.0 1696.0 3571520]
 ['South Atlantic' 'Delaware' 708.0 374.0 965479]
 ['South Atlantic' 'District of Columbia' 3770.0 3134.0 701547]
 ['South Atlantic' 'Florida' 21443.0 9587.0 21244317]
 ['South Atlantic' 'Georgia' 6943.0 2556.0 10511131]
 ['Pacific' 'Hawaii' 4131.0 2399.0 1420593]
 ['Mountain' 'Idaho' 1297.0 715.0 1750536]
 ['East North Central' 'Illinois' 6752.0 3891.0 12723071]
 ['East North Central' 'Indiana' 3776.0 1482.0 6695497]
 ['West North Central' 'Iowa' 1711.0 1038.0 3148618]
 ['West North Central' 'Kansas' 1443.0 773.0 2911359]
 ['East South Central' 'Kentucky' 2735.0 953.0 4461153]
 ['West South Central' 'Louisiana' 2540.0 519.0 4659690]
 ['New 

In [14]:
# Print the column index of homelessness
print(homelessness.columns)

Index(['region', 'state', 'individuals', 'family_members', 'state_pop'], dtype='object')


In [15]:
# Print the row index of homelessness
print(homelessness.index)

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
            50],
           dtype='int64')


A `DataFrame` has three components: values, a column index, and a row index  
  
**Sorting and subsetting**  
  
Sorting - change the order of the row by sorting them.  
  
`df.sort_values("col_name")` a method that takes column_name as argument  
`df.sort_values(["col_name_1", "col_name_2"], ascending = [True, False])` arguments for multiple columns and sorting order
`df["col_name"]`, to look at only one column
`df[["col_name_1"],["col_name_2"]]`, the outer squared brackets are responsible for subsetting the data frame, the inner squared brackets are responsible for creating a list of column names to subset.  
`cols_to_subset = ["col_name_1", "col_name_2"]` , create the list of columns as a variable   
`df[cols_to_subset]`, use the separate list as a variable to subset the df  
  
Subsetting rows - by creating logic to filter by  
`df[df["col_name_1"] > 50]` , subset the df with a logic of a column. We can sub-set also with text and dates  
  
Subsetting using multiple conditions using operators:  
`is_yes = df["col_name_1"] == "yes"`  
`is_no = df["col_name_2"] == "no"`  
`df[is_yes & is_no]`  
  
Or as one line of code, need paranthesis around each condition:  
`df[ (df["col_name_1"] == "yes") & (df["col_name_1"] == "no") ]`  
  
Subsetting using categorical values using `isin()` method:  
`is_black_or_brown = df["color"].isin(["Black", "Brown"])`  
`df[is_black_or_brown]`  

**Sorting rows**  
Examples

In [16]:
homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

Sort `homelessness` by the number of homeless `individuals` in ascending order, and save this as `homelessness_ind`.

In [17]:
# Sort homelessness by individual
homelessness_ind = homelessness.sort_values(["individuals"])

# Print the top few rows
print(homelessness_ind.head())

                region         state  individuals  family_members  state_pop
50            Mountain       Wyoming        434.0           205.0     577601
34  West North Central  North Dakota        467.0            75.0     758080
7       South Atlantic      Delaware        708.0           374.0     965479
39         New England  Rhode Island        747.0           354.0    1058287
45         New England       Vermont        780.0           511.0     624358


Sort `homelessness` by the number of homeless `family_members` in descending order, and save this as `homelessness_fam`.

In [18]:
# Sort homelessness by descending family members
homelessness_fam = homelessness.sort_values(["family_members"], ascending = False)

# Print the top few rows
print(homelessness_fam.head())

                region          state  individuals  family_members  state_pop
32        Mid-Atlantic       New York      39827.0         52070.0   19530351
4              Pacific     California     109008.0         20964.0   39461588
21         New England  Massachusetts       6811.0         13257.0    6882635
9       South Atlantic        Florida      21443.0          9587.0   21244317
43  West South Central          Texas      19199.0          6111.0   28628666


Sort `homelessness` first by region (ascending), and then by number of family members (descending). Save this as `homelessness_reg_fam`.

In [19]:
# Sort homelessness by region, then descending family members
homelessness_reg_fam = homelessness.sort_values(["region", "family_members"], ascending = [True, False])

# Print the top few rows
print(homelessness_reg_fam.head())

                region      state  individuals  family_members  state_pop
13  East North Central   Illinois       6752.0          3891.0   12723071
35  East North Central       Ohio       6929.0          3320.0   11676341
22  East North Central   Michigan       5209.0          3142.0    9984072
49  East North Central  Wisconsin       2740.0          2167.0    5807406
14  East North Central    Indiana       3776.0          1482.0    6695497


Using `.sort_values()` and `.head()` is a powerful pair.

**Subsetting columns**  
When working with data, you may not need all of the variables in your dataset. Square brackets (`[]`) can be used to select only the columns that matter to you in an order that makes sense to you. To select only `"col_a"` of the DataFrame `df`, use  
`df["col_a"]`  
To select two columns use  
`df[["col_a","col_b"]]`  

In [20]:
# Select the individuals column
individuals = homelessness["individuals"]

# Print the head of the result
print(individuals.head())

0      2570.0
1      1434.0
2      7259.0
3      2280.0
4    109008.0
Name: individuals, dtype: float64


In [21]:
# Select the state and family_members columns
state_fam = homelessness[["state","family_members"]]

# Print the head of the result
print(state_fam.head())

        state  family_members
0     Alabama           864.0
1      Alaska           582.0
2     Arizona          2606.0
3    Arkansas           432.0
4  California         20964.0


In [22]:
# Select only the individuals and state columns, in that order
ind_state = homelessness[["individuals","state"]]

# Print the head of the result
print(ind_state.head())

   individuals       state
0       2570.0     Alabama
1       1434.0      Alaska
2       7259.0     Arizona
3       2280.0    Arkansas
4     109008.0  California


**Subsetting rows**  
A large part of data science is about finding which bits of your dataset are interesting. One of the simplest techniques for this is to find a subset of rows that match some criteria. This is sometimes known as * filtering rows*  or *selecting rows*.  
  
There are many ways to subset a DataFrame, perhaps the most common is to use relational operators to return `True` or `False` for each row, then pass that inside square brackets.  
  
`df[df["height_cm"] > 60]`  
`df[df["color"] == "tan"]`  
  
We can filter for multiple conditions at once by using the "bitwise and" operator, `&`.  
  
`df[(df["height_cm"] > 60) & (df["color"] == "tan")]`

In [23]:
# Filter for rows where individuals is greater than 10000
ind_gt_10k = homelessness[homelessness["individuals"] > 10000]

# See the result
print(ind_gt_10k)

                region       state  individuals  family_members  state_pop
4              Pacific  California     109008.0         20964.0   39461588
9       South Atlantic     Florida      21443.0          9587.0   21244317
32        Mid-Atlantic    New York      39827.0         52070.0   19530351
37             Pacific      Oregon      11139.0          3337.0    4181886
43  West South Central       Texas      19199.0          6111.0   28628666
47             Pacific  Washington      16424.0          5880.0    7523869


In [24]:
# Filter for rows where region is Mountain
mountain_reg = homelessness[homelessness["region"] == "Mountain"]

# See the result
print(mountain_reg)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
5   Mountain    Colorado       7607.0          3250.0    5691287
12  Mountain       Idaho       1297.0           715.0    1750536
26  Mountain     Montana        983.0           422.0    1060665
28  Mountain      Nevada       7058.0           486.0    3027341
31  Mountain  New Mexico       1949.0           602.0    2092741
44  Mountain        Utah       1904.0           972.0    3153550
50  Mountain     Wyoming        434.0           205.0     577601


In [25]:
# Filter for rows where family_members is less than 1000 
# and region is Pacific
fam_lt_1k_pac = homelessness[ (homelessness["family_members"] < 1000) & (homelessness["region"]=="Pacific") ]

# See the result
print(fam_lt_1k_pac)

    region   state  individuals  family_members  state_pop
1  Pacific  Alaska       1434.0           582.0     735139


Using square brackets plus logical conditions is often the most powerful way of identifying interesting rows of data.

**Subsetting rows by categorical variables**  
Subsetting data based on a categorical variable often involves using the "or" operator (`|`) to select rows from multiple categories. This can get tedious when you want all states in one of three different regions, for example. Instead, use the `.isin()` method, which will allow you to tackle this problem by writing one condition instead of three separate ones.  
  
`colors = ["brown", "black", "tan"]`  
`condition = df["color"].isin(colors)`  
`df[condition]`

In [26]:
homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [27]:
# Subset for rows in South Atlantic or Mid-Atlantic regions
south_mid_atlantic = homelessness[homelessness["region"].isin(["South Atlantic","Mid-Atlantic"])]

# See the result
print(south_mid_atlantic)

            region                 state  individuals  family_members  \
7   South Atlantic              Delaware        708.0           374.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
9   South Atlantic               Florida      21443.0          9587.0   
10  South Atlantic               Georgia       6943.0          2556.0   
20  South Atlantic              Maryland       4914.0          2230.0   
30    Mid-Atlantic            New Jersey       6048.0          3350.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
33  South Atlantic        North Carolina       6451.0          2817.0   
38    Mid-Atlantic          Pennsylvania       8163.0          5349.0   
40  South Atlantic        South Carolina       3082.0           851.0   
46  South Atlantic              Virginia       3928.0          2047.0   
48  South Atlantic         West Virginia       1021.0           222.0   

    state_pop  
7      965479  
8      701547  
9 

In [28]:
# The Mojave Desert states
canu = ["California", "Arizona", "Nevada", "Utah"]

# Filter for rows in the Mojave Desert states
mojave_homelessness = homelessness[homelessness["state"].isin(canu)]

# See the result
print(mojave_homelessness)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
4    Pacific  California     109008.0         20964.0   39461588
28  Mountain      Nevada       7058.0           486.0    3027341
44  Mountain        Utah       1904.0           972.0    3153550


Using `.isin()` makes subsetting categorical variables a breeze  
  
**New Columns**  
  
Often we want to add new columns derived from existing columns. Adding a new column can go by many names like mutating a data frame, transforming a data frame or feature engineering.  
  
Example of how to add a new column from existing columns:  
`df["heigh_m"] = df["height_cm"]  / 100`  
On the left hand side we have in square brackets the name of the column we want to create.  
On the right hand side we have the calculation.  
We will then have both the existing and the new column in the data frame.  
  
**Multiple manipulations**  
Skinny dogs e.g  
`bmi_lt_100 = df[df["bmi] < 100]`  
`bmi_lt_100_height = bmi_lt_100.sort_values("height_cm", ascending=False)`  
`bmi_lt_100_height[["name", "height_cm", "bmi"]]`  
  
Some examples:

**Adding new columns**  
You aren't stuck with just the data you are given. Instead, you can add new columns to a DataFrame. This has many names, such as *transforming*, *mutating*, and *feature engineering*.  
  
You can create new columns from scratch, but it is also common to derive them from other columns, for example, by adding columns together or by changing their units.

In [29]:
# Add total col as sum of individuals and family_members
homelessness["total"] = homelessness["individuals"] + homelessness["family_members"]

# Add p_individuals col as proportion of individuals
homelessness["p_individuals"] = homelessness["individuals"] / homelessness["total"]

# See the result
print(homelessness.head())

               region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  
0    3434.0       0.748398  
1    2016.0       0.711310  
2    9865.0       0.735834  
3    2712.0       0.840708  
4  129972.0       0.838704  


If our dataset doesn't have the exact columns you need, we can often make your own from what you have.  
  
**Combo-attack!**  
You've seen the four most common types of data manipulation: sorting rows, subsetting columns, subsetting rows, and adding new columns. In a real-life data analysis, you can mix and match these four manipulations to answer a multitude of questions.  
  
In this exercise, you'll answer the question, "Which state has the highest number of homeless individuals per 10,000 people in the state?"  

In [30]:
# Create indiv_per_10k col as homeless individuals per 10k state pop
homelessness["indiv_per_10k"] = 10000 * homelessness["individuals"] / homelessness["state_pop"] 

# Subset rows for indiv_per_10k greater than 20
high_homelessness = homelessness[homelessness["indiv_per_10k"] > 20]

# Sort high_homelessness by descending indiv_per_10k
high_homelessness_srt = high_homelessness.sort_values("indiv_per_10k", ascending=False) 

# From high_homelessness_srt, select the state and indiv_per_10k cols
result = high_homelessness_srt[["state","indiv_per_10k"]]

# See the result
print(result)

                   state  indiv_per_10k
8   District of Columbia      53.738381
11                Hawaii      29.079406
4             California      27.623825
37                Oregon      26.636307
28                Nevada      23.314189
47            Washington      21.829195
32              New York      20.392363


Washington, D.C. has the highest number of homeless individuals - almost 54 per ten thousand people. This is almost double the number of the next-highest state, Hawaii. If you combine new column addition, row subsetting, sorting, and column selection, you can answer lots of questions like this.  
  
### 2. Aggregating DataFrames  
  
**Summary statistics**  
Focus here is on aggregating data using summary statistics. We will use dot-notation. E.g.  
`df["col_1"].mean()`  
There are many more summary statistics that we can calculate like:  
`.median()`  
`.mode()`  
`.min()`  
`.max()`  
`.var()`  
`.std()`  
We can also take sums and calculate quantiles:  
`.sum()`  
`.quantile()`  
  
The .agg() methods  
  
Allows for calculating custom summary statistics.  
`def pct30(column)`  
` return column.quantile(0.3)`  
Calculate the 30th percentile of df columns  
`df["col_1"].agg(pct30)` 
  
Summaries on multiple columns  
`df[["col_1", "col_2"]].agg(pct30)`  
  
Multiple summaries  
`def pct40(column)`  
` return column.quantile(0.4)`  
Calculate the 30th and 40th percentile of df column  
`df["col_1"].agg([pct30, pct40])`  
  
Cumulative sum  
`df["col_1].cumsum()`
It returns one sum for each row of a df.  
Example see below:

In [3]:
import pandas as pd

# OMA Need to create the DF
path = "data/homelessness.csv"
homelessness = pd.read_csv(path, index_col=0)

homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [4]:
# cumsum
homelessness["ind_cumsum"] = homelessness["individuals"].cumsum()
homelessness[["individuals","ind_cumsum"]].head()

individuals  ind_cumsum
0       2570.0      2570.0
1       1434.0      4004.0
2       7259.0     11263.0
3       2280.0     13543.0
4     109008.0    122551.0

Other Cumulative Statistics  
`.cummax()`  
`.cummin()`  
`.cumprod()`  
All these return an entire column rather than one value.  
  
**Mean and median**  
Some examples

In [5]:
# OMA need to read in the data first
path2 = "data/sales_subset.csv"
sales = pd.read_csv(path2, index_col=0)
sales.head()

store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

In [6]:
# Print the head of the sales DataFrame
print(sales.head())

# Print the info about the sales DataFrame
print(sales.info())

# Print the mean of weekly_sales
print(sales["weekly_sales"].mean())

# Print the median of weekly_sales
print(sales["weekly_sales"].median())

   store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808  
<class 'pandas.core.frame.DataFrame'>
Int64Index: 10774 entries, 0 to 10773
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   store                 10774 non-null  int64  
 1   t

The mean weekly sales amount is almost double the median weekly sales amount! This can tell you that there are a few very high sales weeks that are making the mean so much higher than the median.  
  
**Summarizing dates**  
Summary statistics can also be calculated on date columns that have values with the data type `datetime64`. Some summary statistics — like mean — don't make a ton of sense on dates, but others are super helpful, for example, minimum and maximum, which allow you to see what time range your data covers.

In [7]:
# Print the maximum of the date column
print(sales["date"].max())

# Print the minimum of the date column
print(sales["date"].min())

2012-10-26
2010-02-05


Taking the minimum and maximum of a column of dates is handy for figuring out what time period your data covers. In this case, there are data from February of 2010 to October of 2012.  
  
**Efficient summaries**  
While pandas and NumPy have tons of functions, sometimes, you may need a different function to summarize your data.  
  
The `.agg()` method allows you to apply your own custom functions to a DataFrame, as well as apply functions to more than one column of a DataFrame at once, making your aggregations super-efficient. For example,  
`df['column'].agg(function)`  

In [8]:
# A custom IQR function
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)
    
# Print IQR of the temperature_c column
print(sales["temperature_c"].agg(iqr))

16.58333333333334


In [9]:
# A custom IQR function
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)

# Update to print IQR of temperature_c, fuel_price_usd_per_l, & unemployment
print(sales[["temperature_c", "fuel_price_usd_per_l", "unemployment"]].agg(iqr))

temperature_c           16.583333
fuel_price_usd_per_l     0.073176
unemployment             0.565000
dtype: float64


In [10]:
# Import NumPy and create custom IQR function
import numpy as np
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)

# Update to print IQR and median of temperature_c, fuel_price_usd_per_l, & unemployment
print(sales[["temperature_c", "fuel_price_usd_per_l", "unemployment"]].agg([iqr,np.median]))

        temperature_c  fuel_price_usd_per_l  unemployment
iqr         16.583333              0.073176         0.565
median      16.966667              0.743381         8.099


The `.agg()` method makes it easy to compute multiple statistics on multiple columns, all in just one line of code.  
  
**Cumulative statistics**  
Cumulative statistics can also be helpful in tracking summary statistics over time. In this exercise, you'll calculate the cumulative sum and cumulative max of a department's weekly sales, which will allow you to identify what the total sales were so far as well as what the highest weekly sales were so far.

In [24]:
# Create dataset needed. Only include store = 1 and department = 1
sales_1_1 = sales[ (sales["store"]==1) & (sales["department"]==1) ]
sales_1_1.shape

(144, 9)

In [25]:
sales_1_1.head()

store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

In [27]:
# Sort sales_1_1 by date
sales_1_1 = sales_1_1.sort_values("date").head(12)

# Get the cumulative sum of weekly_sales, add as cum_weekly_sales col
sales_1_1["cum_weekly_sales"] = sales_1_1["weekly_sales"].cumsum()

# Get the cumulative max of weekly_sales, add as cum_max_sales col
sales_1_1["cum_max_sales"] = sales_1_1["weekly_sales"].cummax()

# See the columns you calculated
print(sales_1_1[["date", "weekly_sales", "cum_weekly_sales", "cum_max_sales"]])

            date  weekly_sales  cum_weekly_sales  cum_max_sales
0     2010-02-05      24924.50          24924.50       24924.50
9009  2010-02-05      18187.71          43112.21       24924.50
8109  2010-02-05      32313.79          75426.00       32313.79
7199  2010-02-05      46021.21         121447.21       46021.21
6293  2010-02-05      21500.58         142947.79       46021.21
5408  2010-02-05      32842.31         175790.10       46021.21
4495  2010-02-05      46761.90         222552.00       46761.90
3593  2010-02-05      40212.84         262764.84       46761.90
2699  2010-02-05      25619.00         288383.84       46761.90
1798  2010-02-05      38724.42         327108.26       46761.90
901   2010-02-05      35034.06         362142.32       46761.90
9899  2010-02-05      21244.50         383386.82       46761.90


Not all functions that calculate on columns return a single number. Some, like the cumulative statistic functions, return a whole column.  
  
**Counting**  
  
Now we will summarize categorical data using counting.  
  
Doprring duplicate names  
`df.drop_duplicates(subset="col_name")`  
  
This method takes an argument, the column_name, we want to find our duplicates based on. In this case we want all the unique names. We can drop based on two columns.  
`df.drop_duplicates(subset=["col_name_1","col_name_2])`  
  
Now we base our duplicate dropping on multiple columns by passing a list of column names into the `subset` argument. Now for example if we have string `Tom` twice in the `col_name_1` but with `Wattson` and `Johnson` in `col_name_2` we can thus keep duplicates in the `col_name_1` since it is the combination of both columns that needs to be unique.  
  
To count the rows in each `col_name_1` we can  
`df["col_name_1"].value_counts(sort=True)` subset the df by col_name_1 and count the rows and sort the output. We can also use the `normalize` argument to turn the counts into proportions of total.  
`df["col_name_1"].value_counts(normalize=True)`  
  
Some exaxmples.  
  
**Dropping duplicates**  
Removing duplicates is an essential skill to get accurate counts because often, you don't want to count the same thing multiple times. In this exercise, you'll create some new DataFrames using unique values from `sales`.

In [30]:
# Drop duplicate store/type combinations
store_types = sales.drop_duplicates(subset=["store", "type"])
print(store_types.head())

# Drop duplicate store/department combinations
store_depts = sales.drop_duplicates(subset=["store", "department"])
print(store_depts.head())

# Subset the rows where is_holiday is True and drop duplicate dates
holiday_dates = sales[sales["is_holiday"]].drop_duplicates(subset="date")

# Print date col of holiday_dates
print(holiday_dates["date"])

      store type  department        date  weekly_sales  is_holiday  \
0         1    A           1  2010-02-05      24924.50       False   
3593      1    B           1  2010-02-05      40212.84       False   

      temperature_c  fuel_price_usd_per_l  unemployment  
0          5.727778              0.679451         8.106  
3593      12.411111              0.782478         9.765  
    store type  department        date  weekly_sales  is_holiday  \
0       1    A           1  2010-02-05      24924.50       False   
12      1    A           2  2010-02-05      50605.27       False   
24      1    A           3  2010-02-05      13740.12       False   
36      1    A           4  2010-02-05      39954.04       False   
48      1    A           5  2010-02-05      32229.38       False   

    temperature_c  fuel_price_usd_per_l  unemployment  
0        5.727778              0.679451         8.106  
12       5.727778              0.679451         8.106  
24       5.727778              0.67945

The holiday weeks correspond to the Superbowl in February, Labor Day in September, Thanksgiving in November, and Christmas in December. Now that the duplicates are removed, it's time to do some counting.  
  
**Counting categorical variables**  
Counting is a great way to get an overview of your data and to spot curiosities that you might not notice otherwise. In this exercise, you'll count the number of each type of store and the number of each department number using the DataFrames you created in the previous exercise:  
  
`# Drop duplicate store/type combinations`  
`store_types = sales.drop_duplicates(subset=["store", "type"])`  
  
`# Drop duplicate store/department combinations`  
`store_depts = sales.drop_duplicates(subset=["store", "department"])`

In [32]:
# Count the number of stores of each type
store_counts = store_types["type"].value_counts()
print(store_counts)

# Get the proportion of stores of each type
store_props = store_types["type"].value_counts(normalize=True)
print(store_props)

# Count the number of each department number and sort
dept_counts_sorted = store_depts["department"].value_counts()
print(dept_counts_sorted)

# Get the proportion of departments of each number and sort
dept_props_sorted = store_depts["department"].value_counts(sort=True, normalize=True)
print(dept_props_sorted)

B    1
A    1
Name: type, dtype: int64
B    0.5
A    0.5
Name: type, dtype: float64
99    1
98    1
23    1
24    1
25    1
     ..
58    1
59    1
60    1
67    1
1     1
Name: department, Length: 80, dtype: int64
99    0.0125
98    0.0125
23    0.0125
24    0.0125
25    0.0125
       ...  
58    0.0125
59    0.0125
60    0.0125
67    0.0125
1     0.0125
Name: department, Length: 80, dtype: float64


**Grouped summary statistics**  
So far we have calculated summary statistics for all rows in a dataset.  
Now we will use summary statistics to compare different groups. We can get more insights of individual groups compared to complete columns.  
  
For example does the weekly_sales depend on store?

In [23]:
sales.head()

store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

In [18]:
print(sales[sales["store"] == 1]["weekly_sales"].mean())
print(sales[sales["store"] == 2]["weekly_sales"].mean())

20896.941786903433
26517.43516164994


We can subset the stores by store name and copy paste the code but that is not very efficient.  
Instead we can use `.groupby()` method and add the column name as argument and the column we want to use for creating the metric `mean()`. This is just one line of code and no copy-paste.

In [19]:
sales.groupby("store")["weekly_sales"].mean()

store
1     20896.941787
2     26517.435162
4     26126.986071
6     21561.186477
10    25696.678370
13    25664.149474
14    30384.003017
19    19930.838157
20    28382.766385
27    24207.474711
31    18178.932225
39    18414.938423
Name: weekly_sales, dtype: float64

We can use the `.agg()` method (just like with ungrouped statistics) to add multiple statistics as a list of arguments.

In [21]:
sales.groupby("store")["weekly_sales"].agg([min,max,sum])

min        max          sum
store                                
1      -698.0  140504.41  18828144.55
2     -1098.0  178982.89  23786139.34
4       -88.0  165765.50  23540414.45
6      -698.0  119812.38  19275700.71
10     -798.0  232558.51  23178403.89
13      -98.0  166872.27  23431368.47
14     -498.0  293966.05  26889842.67
19     -449.0  147449.06  18057339.37
20     -598.0  195223.84  25828317.41
27     -409.0  177605.86  21786727.24
31     -224.0  131062.82  16179249.68
39     -159.0  114691.10  16113071.12

We can also group by multiple columns and calculate summary statistics.  
E.g group by `["store","is_holiday"]` and we take the `mean()` of two columns `["weekly_sales","fuel_price_usd_per_l"]`

In [26]:
sales.groupby(["store","is_holiday"])[["weekly_sales","fuel_price_usd_per_l"]].mean()

weekly_sales  fuel_price_usd_per_l
store is_holiday                                    
1     False       20962.474477              0.716806
      True         1280.823333              0.795775
2     False       26601.621409              0.715482
      True         1429.933333              0.795775
4     False       26270.158984              0.718595
      True          470.400000              0.714269
6     False       21606.366267              0.716549
      True         1411.000000              0.854861
10    False       25751.980533              0.805676
      True          810.705000              0.887751
13    False       25719.082843              0.749545
      True          642.000000              0.807839
14    False       30486.986939              0.764505
      True          106.730000              0.789346
19    False       20107.698263              0.803097
      True           78.291250              0.907134
20    False       28507.329338              0.766184
      True          169.257500              0.850304
27    False       24308.986250              0.802730
      True         1468.890000              0.875599
31    False       18218.964200              0.715714
      True          404.735000              0.766231
39    False       18498.700528              0.717155
      True          175.740000              0.718548

Practise!  

In [27]:
# Calc total weekly sales
sales_all = sales["weekly_sales"].sum()

# Subset for type A stores, calc total weekly sales
sales_A = sales[sales["type"] == "A"]["weekly_sales"].sum()

# Subset for type B stores, calc total weekly sales
sales_B = sales[sales["type"] == "B"]["weekly_sales"].sum()

# Subset for type C stores, calc total weekly sales
sales_C = sales[sales["type"] == "C"]["weekly_sales"].sum()

# Get proportion for each type
sales_propn_by_type = [sales_A, sales_B, sales_C] / sales_all
print(sales_propn_by_type)

[0.9097747 0.0902253 0.       ]


About 91% of sales occurred in stores of type A', 9% in stores of type B, and there are no sales records for stores of type C. Now see if we can do this calculation using `.groupby()`.

In [28]:
# Group by type; calc total weekly sales
sales_by_type = sales.groupby("type")["weekly_sales"].sum()

# Get proportion for each type
sales_propn_by_type = sales_by_type / sum(sales["weekly_sales"])
print(sales_propn_by_type)

type
A    0.909775
B    0.090225
Name: weekly_sales, dtype: float64


In [32]:
# From previous step
sales_by_type = sales.groupby("type")["weekly_sales"].sum()

# Group by type and is_holiday; calc total weekly sales
sales_by_type_is_holiday = sales.groupby(["type","is_holiday"])["weekly_sales"].sum()
print(sales_by_type_is_holiday)

type  is_holiday
A     False         2.336927e+08
      True          2.360181e+04
B     False         2.317678e+07
      True          1.621410e+03
Name: weekly_sales, dtype: float64


We were able to do the same calculation as in the previous exercise while writing much less code.  
  
**Multiple grouped summaries**  
Earlier in this chapter, you saw that the `.agg()` method is useful to compute multiple statistics on multiple variables. It also works with grouped data. NumPy, which is imported as `np`, has many different summary statistics functions, including: `np.min`, `np.max`, `np.mean,` and `np.median`.

In [33]:
# Import numpy with the alias np
import numpy as np

# For each store type, aggregate weekly_sales: get min, max, mean, and median
sales_stats = sales.groupby("type")["weekly_sales"].agg([np.min,np.max,np.mean,np.median])

# Print sales_stats
print(sales_stats)

# For each store type, aggregate unemployment and fuel_price_usd_per_l: get min, max, mean, and median
unemp_fuel_stats = sales.groupby("type")[["unemployment","fuel_price_usd_per_l"]].agg([np.min,np.max,np.mean,np.median])

# Print unemp_fuel_stats
print(unemp_fuel_stats)

        amin       amax          mean    median
type                                           
A    -1098.0  293966.05  23674.667242  11943.92
B     -798.0  232558.51  25696.678370  13336.08
     unemployment                         fuel_price_usd_per_l            \
             amin   amax      mean median                 amin      amax   
type                                                                       
A           3.879  8.992  7.972611  8.067             0.664129  1.107410   
B           7.170  9.765  9.279323  9.199             0.760023  1.107674   

                          
          mean    median  
type                      
A     0.744619  0.735455  
B     0.805858  0.803348  


Notice that the minimum `weekly_sales` is negative because some stores had more returns than sales.  
  
**Pivot tables**  
Pivot tables are another way of calculating grouped summary statistics. We can create pivot tables in pandas, simlar to spread sheets.  
  
Group by to pivot table. Previously we calculated `mean()` of `weekly_sales` grouped by `"store"`.  
We can do the same thing using `pivot_table()` method.

In [34]:
# as per before
sales.groupby("store")["weekly_sales"].mean()

store
1     20896.941787
2     26517.435162
4     26126.986071
6     21561.186477
10    25696.678370
13    25664.149474
14    30384.003017
19    19930.838157
20    28382.766385
27    24207.474711
31    18178.932225
39    18414.938423
Name: weekly_sales, dtype: float64

In [35]:
# using pivot table
sales.pivot_table(values = "weekly_sales",
                  index = "store")

weekly_sales
store              
1      20896.941787
2      26517.435162
4      26126.986071
6      21561.186477
10     25696.678370
13     25664.149474
14     30384.003017
19     19930.838157
20     28382.766385
27     24207.474711
31     18178.932225
39     18414.938423

The `values` argument is the column that we want to summarise.  
The  `index` argument is the column that we want to group by.  
By default `.pivot_table()` method takes the `mean()` value for each group.  
  
**Different statistics**  
For different statistics using `.pivot_table()` we can use the `aggfunc` argument and pass it a funciton. E.g.

In [36]:
# single statistics
import numpy as np
sales.pivot_table(values = "weekly_sales",
                  index = "store",
                  aggfunc = np.median)

weekly_sales
store              
1          9775.170
2         13764.940
4         13063.930
6         13200.780
10        13336.080
13        13050.060
14        14792.990
19        11091.770
20        14706.675
27        12689.850
31         7787.725
39         8980.720

In [37]:
# multiple statistics
sales.pivot_table(values = "weekly_sales",
                  index = "store",
                  aggfunc = [np.mean,np.median])

mean       median
       weekly_sales weekly_sales
store                           
1      20896.941787     9775.170
2      26517.435162    13764.940
4      26126.986071    13063.930
6      21561.186477    13200.780
10     25696.678370    13336.080
13     25664.149474    13050.060
14     30384.003017    14792.990
19     19930.838157    11091.770
20     28382.766385    14706.675
27     24207.474711    12689.850
31     18178.932225     7787.725
39     18414.938423     8980.720

**Pivot on two variables**  
We can calculate the the `mean()` on two columns as we did before using `pivot_table()` method.  
We pass the second variable into the `columns` argument in such case.

In [38]:
# from before using group_by
sales.groupby(["type","is_holiday"])["weekly_sales"].mean()

type  is_holiday
A     False         23768.583523
      True            590.045250
B     False         25751.980533
      True            810.705000
Name: weekly_sales, dtype: float64

In [39]:
# using pivot_table method
sales.pivot_table(values = "weekly_sales", index="type", columns="is_holiday")

is_holiday         False      True 
type                               
A           23768.583523  590.04525
B           25751.980533  810.70500

To bet rid of any **missing values** we can use the `fill_value=0` argument in our `.pivot_table()`. If we set the `margins = True` then the last row contains the mean values of the rows in each column and the last column will contain the mean of all the columns in each row. The last value on the right corner is the mean value of all store types regardless of holiday.

In [43]:
sales.pivot_table(values = "weekly_sales", index="type", columns="is_holiday", fill_value=0, margins=True)

is_holiday         False        True           All
type                                              
A           23768.583523  590.045250  23674.667242
B           25751.980533  810.705000  25696.678370
All         23934.913873  600.552857  23843.950149

Pivot tables are the standard way of aggregating data in spreadsheets. In pandas, pivot tables are essentially just another way of performing grouped calculations. That is, the `.pivot_table()` method is just an alternative to `.groupby()`.  
  
Some examples:

In [44]:
# Pivot for mean weekly_sales for each store type. Takes mean() by default!
mean_sales_by_type = sales.pivot_table(values="weekly_sales", index="type")

# Print mean_sales_by_type
print(mean_sales_by_type)

      weekly_sales
type              
A     23674.667242
B     25696.678370


In [45]:
# Import NumPy as np
import numpy as np

# Pivot for mean and median weekly_sales for each store type
mean_med_sales_by_type = sales.pivot_table(values = "weekly_sales", index = "type", aggfunc=[np.mean,np.median])

# Print mean_med_sales_by_type
print(mean_med_sales_by_type)

              mean       median
      weekly_sales weekly_sales
type                           
A     23674.667242     11943.92
B     25696.678370     13336.08


In [46]:
# Pivot for mean weekly_sales by store type and holiday 
mean_sales_by_type_holiday = sales.pivot_table(values = "weekly_sales", index = "type", columns = "is_holiday")

# Print mean_sales_by_type_holiday
print(mean_sales_by_type_holiday)

is_holiday         False      True 
type                               
A           23768.583523  590.04525
B           25751.980533  810.70500


In [48]:
# Print mean weekly_sales by department and type; fill missing values with 0
print(sales.pivot_table(values = "weekly_sales", index = "department", columns = "type", fill_value=0))

type                    A              B
department                              
1            30961.725379   44050.626667
2            67600.158788  112958.526667
3            17160.002955   30580.655000
4            44285.399091   51219.654167
5            34821.011364   63236.875000
...                   ...            ...
95          123933.787121   77082.102500
96           21367.042857    9528.538333
97           28471.266970    5828.873333
98           12875.423182     217.428333
99             379.123659       0.000000

[80 rows x 2 columns]


In [49]:
# Print the mean weekly_sales by department and type; fill missing values with 0s; sum all rows and cols
print(sales.pivot_table(values="weekly_sales", index="department", columns="type", fill_value=0, margins=True))

type                   A              B           All
department                                           
1           30961.725379   44050.626667  32052.467153
2           67600.158788  112958.526667  71380.022778
3           17160.002955   30580.655000  18278.390625
4           44285.399091   51219.654167  44863.253681
5           34821.011364   63236.875000  37189.000000
...                  ...            ...           ...
96          21367.042857    9528.538333  20337.607681
97          28471.266970    5828.873333  26584.400833
98          12875.423182     217.428333  11820.590278
99            379.123659       0.000000    379.123659
All         23674.667242   25696.678370  23843.950149

[81 rows x 3 columns]


We are now armed with pivot table skills that can help us compute summaries at multiple grouped levels in one line of code.  
  
### 3. Slicing and Indexing DataFrames  
  
**Explicit Indexes**  
As we saw in ch 1 DataFrames are composed of three parts:  
Numpy array for data and two indexes to store column and row information.  
  
**df.columns**
contains and index of column names  
  
**df.index**  
contains and index of row numbers

In [7]:
import pandas as pd
# OMA need to read in the data first
path2 = "data/sales_subset.csv"
sales = pd.read_csv(path2, index_col=0)

print(sales.columns)

print("\n========================\n")

print(sales.index)

sales.head()

Index(['store', 'type', 'department', 'date', 'weekly_sales', 'is_holiday',
       'temperature_c', 'fuel_price_usd_per_l', 'unemployment'],
      dtype='object')


Int64Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,
                9,
            ...
            10764, 10765, 10766, 10767, 10768, 10769, 10770, 10771, 10772,
            10773],
           dtype='int64', length=10774)


store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

**Setting a column as the index** using the `set_index()` method.  
This will change the output. A note that the column chosen as index is now left aligned rather than right aligned.  
  
**Removing and index** using the `reset_index()` method.  
This undo´s the previously set column as index.  
It is also possible to drop the columed used as index entirely from the data frame by using argument `reset_index(drop=True)`.

In [8]:
# set index
sales_ind = sales.set_index("date")
sales_ind.head()

store type  department  weekly_sales  is_holiday  temperature_c  \
date                                                                          
2010-02-05      1    A           1      24924.50       False       5.727778   
2010-03-05      1    A           1      21827.90       False       8.055556   
2010-04-02      1    A           1      57258.43       False      16.816667   
2010-05-07      1    A           1      17413.94       False      22.527778   
2010-06-04      1    A           1      17558.09       False      27.050000   

            fuel_price_usd_per_l  unemployment  
date                                            
2010-02-05              0.679451         8.106  
2010-03-05              0.693452         8.106  
2010-04-02              0.718284         7.808  
2010-05-07              0.748928         7.808  
2010-06-04              0.714586         7.808

In [11]:
# reset the index back
sales_ind.reset_index().head()

date  store type  department  weekly_sales  is_holiday  \
0  2010-02-05      1    A           1      24924.50       False   
1  2010-03-05      1    A           1      21827.90       False   
2  2010-04-02      1    A           1      57258.43       False   
3  2010-05-07      1    A           1      17413.94       False   
4  2010-06-04      1    A           1      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

In [12]:
# reset the index and drop the column
sales_ind.reset_index(drop=True).head()

store type  department  weekly_sales  is_holiday  temperature_c  \
0      1    A           1      24924.50       False       5.727778   
1      1    A           1      21827.90       False       8.055556   
2      1    A           1      57258.43       False      16.816667   
3      1    A           1      17413.94       False      22.527778   
4      1    A           1      17558.09       False      27.050000   

   fuel_price_usd_per_l  unemployment  
0              0.679451         8.106  
1              0.693452         8.106  
2              0.718284         7.808  
3              0.748928         7.808  
4              0.714586         7.808

**Indexes make subsetting easiers**  
The reason for using indexes is that is makes it simpler to subset a dataframe.  
Two example of subsetting without and with index setting.  
  
`df[df["col"].isin(["value_1","value_2"])]` without index  
`df_ind.loc[["value_1","value_2]]` with index.  
  
`DataFrames` have indexing method `loc()`, in the latter we just pass subsetting values as a `list`. This is easier!  
  
Note that the values in the index do not need to be unique. It can return multiple rows for a value. We can also use multiple colum names as index.  
  
**Multi-level indexes or hierarchical indexes**  
We pass multiple columns as a list to the `set_index(["col_1", "col_2"])` method.  
The second index `"col_1"` is nested inside the outer level index `"col_1"`.  
  
**Subset the outer level of index with a list**  
`df.loc[["col_1", "col_2"]]`  
Example

In [19]:
# set multilevel index of two cols
sales_ind = sales.set_index(["date","type"])

# subset the outer level index with a list
sales_ind.loc[["2010-02-05","2010-03-05"]]

store  department  weekly_sales  is_holiday  temperature_c  \
date       type                                                               
2010-02-05 A         1           1      24924.50       False       5.727778   
           A         1           2      50605.27       False       5.727778   
           A         1           3      13740.12       False       5.727778   
           A         1           4      39954.04       False       5.727778   
           A         1           5      32229.38       False       5.727778   
...                ...         ...           ...         ...            ...   
2010-03-05 A        39          95      90747.72       False      10.516667   
           A        39          96      19295.44       False      10.516667   
           A        39          97      20792.64       False      10.516667   
           A        39          98       9469.87       False      10.516667   
           A        39          99          0.01       False      10.516667   

                 fuel_price_usd_per_l  unemployment  
date       type                                      
2010-02-05 A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
...                               ...           ...  
2010-03-05 A                 0.693452         8.554  
           A                 0.693452         8.554  
           A                 0.693452         8.554  
           A                 0.693452         8.554  
           A                 0.693452         8.554  

[1736 rows x 7 columns]

**Subset inner levels with a list of tuples**  
`df.loc[[("value_1","value_A), ("value_2", "value_B")]]`  
The first tuple specifies the value_1 on the outer level and value_A on the inner level.  
The resulting output must match all the conditions of the tuples.  
Example

In [23]:
sales_ind.loc[[("2010-02-05","A"), ("2010-03-05","B") ]]

store  department  weekly_sales  is_holiday  temperature_c  \
date       type                                                               
2010-02-05 A         1           1      24924.50       False       5.727778   
           A         1           2      50605.27       False       5.727778   
           A         1           3      13740.12       False       5.727778   
           A         1           4      39954.04       False       5.727778   
           A         1           5      32229.38       False       5.727778   
...                ...         ...           ...         ...            ...   
2010-03-05 B        10          94          4.52       False      13.288889   
           B        10          95      74608.89       False      13.288889   
           B        10          96       8679.74       False      13.288889   
           B        10          97       5769.35       False      13.288889   
           B        10          98         11.00       False      13.288889   

                 fuel_price_usd_per_l  unemployment  
date       type                                      
2010-02-05 A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
...                               ...           ...  
2010-03-05 B                 0.760023         9.765  
           B                 0.760023         9.765  
           B                 0.760023         9.765  
           B                 0.760023         9.765  
           B                 0.760023         9.765  

[869 rows x 7 columns]

**Sorting by index values**  using the following method  
`df.sort_index()`  
By default it sorts all rows on outer to inner index in ascending order.  
We can control the sorting by passing arguments like `level` and `ascending` arguments.  
`df.sort_index(level=["value_1", "value_2"], ascending=[True, False])`  
Example

In [24]:
# plain sort
sales_ind.sort_index()

store  department  weekly_sales  is_holiday  temperature_c  \
date       type                                                               
2010-02-05 A         1           1      24924.50       False       5.727778   
           A         1           2      50605.27       False       5.727778   
           A         1           3      13740.12       False       5.727778   
           A         1           4      39954.04       False       5.727778   
           A         1           5      32229.38       False       5.727778   
...                ...         ...           ...         ...            ...   
2012-10-05 A        20          99        660.00       False      15.983333   
           A        27          99        435.00       False      18.055556   
           A        39          99        915.00       False      22.250000   
2012-10-12 A        14          96          3.00       False      12.483333   
2012-10-26 A         6          77        -21.63       False      22.411111   

                 fuel_price_usd_per_l  unemployment  
date       type                                      
2010-02-05 A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
           A                 0.679451         8.106  
...                               ...           ...  
2012-10-05 A                 1.052726         7.293  
           A                 1.096579         8.000  
           A                 0.955511         6.228  
2012-10-12 A                 1.056689         8.667  
2012-10-26 A                 0.926188         5.329  

[10774 rows x 7 columns]

In [25]:
# plain sort
sales_ind.sort_index(level=["date","type"], ascending=[True, False])

store  department  weekly_sales  is_holiday  temperature_c  \
date       type                                                               
2010-02-05 B        10           1      40212.84       False      12.411111   
           B        10           2     123952.48       False      12.411111   
           B        10           3      30175.46       False      12.411111   
           B        10           4      51885.86       False      12.411111   
           B        10           5      75297.91       False      12.411111   
...                ...         ...           ...         ...            ...   
2012-10-05 A        20          99        660.00       False      15.983333   
           A        27          99        435.00       False      18.055556   
           A        39          99        915.00       False      22.250000   
2012-10-12 A        14          96          3.00       False      12.483333   
2012-10-26 A         6          77        -21.63       False      22.411111   

                 fuel_price_usd_per_l  unemployment  
date       type                                      
2010-02-05 B                 0.782478         9.765  
           B                 0.782478         9.765  
           B                 0.782478         9.765  
           B                 0.782478         9.765  
           B                 0.782478         9.765  
...                               ...           ...  
2012-10-05 A                 1.052726         7.293  
           A                 1.096579         8.000  
           A                 0.955511         6.228  
2012-10-12 A                 1.056689         8.667  
2012-10-26 A                 0.926188         5.329  

[10774 rows x 7 columns]

**Now we have two problems** with indexex  
* Index values are just data
* Indexes violate "tidy data" principles, since index values dont get their own columns  
* In pandas the syntax of working with indexes is different from syntax of working with columns. This means we have two syntaxes. We can decide not to use indexes. However, it is usefull to know how they work if we need to work with someone elses code.  
  
More examples:

In [26]:
import pandas as pd
# OMA need to read in the data first
path3 = "data/temperatures.csv"
temperatures = pd.read_csv(path3, index_col=0)
print(temperatures.shape)
temperatures.head()

(16500, 4)


date     city        country  avg_temp_c
0  2000-01-01  Abidjan  Côte D'Ivoire      27.293
1  2000-02-01  Abidjan  Côte D'Ivoire      27.685
2  2000-03-01  Abidjan  Côte D'Ivoire      29.061
3  2000-04-01  Abidjan  Côte D'Ivoire      28.162
4  2000-05-01  Abidjan  Côte D'Ivoire      27.547

In [27]:
# Index temperatures by city
temperatures_ind = temperatures.set_index("city")

# Look at temperatures_ind
print(temperatures_ind)

# Reset the index, keeping its contents
print(temperatures_ind.reset_index())

# Reset the index, dropping its contents
print(temperatures_ind.reset_index(drop=True))

               date        country  avg_temp_c
city                                          
Abidjan  2000-01-01  Côte D'Ivoire      27.293
Abidjan  2000-02-01  Côte D'Ivoire      27.685
Abidjan  2000-03-01  Côte D'Ivoire      29.061
Abidjan  2000-04-01  Côte D'Ivoire      28.162
Abidjan  2000-05-01  Côte D'Ivoire      27.547
...             ...            ...         ...
Xian     2013-05-01          China      18.979
Xian     2013-06-01          China      23.522
Xian     2013-07-01          China      25.251
Xian     2013-08-01          China      24.528
Xian     2013-09-01          China         NaN

[16500 rows x 3 columns]
          city        date        country  avg_temp_c
0      Abidjan  2000-01-01  Côte D'Ivoire      27.293
1      Abidjan  2000-02-01  Côte D'Ivoire      27.685
2      Abidjan  2000-03-01  Côte D'Ivoire      29.061
3      Abidjan  2000-04-01  Côte D'Ivoire      28.162
4      Abidjan  2000-05-01  Côte D'Ivoire      27.547
...        ...         ...            .

Setting an index allows more concise code for subsetting for rows of a categorical variable via `.loc[].`  
  
More example:

In [28]:
# Make a list of cities to subset on
cities = ["Moscow", "Saint Petersburg"]

# Subset temperatures using square brackets
print(temperatures[temperatures["city"].isin(cities)])

# Subset temperatures_ind using .loc[]
print(temperatures_ind.loc[cities])

             date              city country  avg_temp_c
10725  2000-01-01            Moscow  Russia      -7.313
10726  2000-02-01            Moscow  Russia      -3.551
10727  2000-03-01            Moscow  Russia      -1.661
10728  2000-04-01            Moscow  Russia      10.096
10729  2000-05-01            Moscow  Russia      10.357
...           ...               ...     ...         ...
13360  2013-05-01  Saint Petersburg  Russia      12.355
13361  2013-06-01  Saint Petersburg  Russia      17.185
13362  2013-07-01  Saint Petersburg  Russia      17.234
13363  2013-08-01  Saint Petersburg  Russia      17.153
13364  2013-09-01  Saint Petersburg  Russia         NaN

[330 rows x 4 columns]
                        date country  avg_temp_c
city                                            
Moscow            2000-01-01  Russia      -7.313
Moscow            2000-02-01  Russia      -3.551
Moscow            2000-03-01  Russia      -1.661
Moscow            2000-04-01  Russia      10.096
Moscow    

`.loc[]` is used by all the best folk! Setting an index allows more concise code for subsetting rows via `.loc[]`.  
  
More example using multi-level indexes.

In [29]:
# Index temperatures by country & city
temperatures_ind = temperatures.set_index(["country","city"])

# List of tuples: Brazil, Rio De Janeiro & Pakistan, Lahore
rows_to_keep = [("Brazil", "Rio De Janeiro"), ("Pakistan","Lahore")]

# Subset for rows to keep
print(temperatures_ind.loc[rows_to_keep])

                               date  avg_temp_c
country  city                                  
Brazil   Rio De Janeiro  2000-01-01      25.974
         Rio De Janeiro  2000-02-01      26.699
         Rio De Janeiro  2000-03-01      26.270
         Rio De Janeiro  2000-04-01      25.750
         Rio De Janeiro  2000-05-01      24.356
...                             ...         ...
Pakistan Lahore          2013-05-01      33.457
         Lahore          2013-06-01      34.456
         Lahore          2013-07-01      33.279
         Lahore          2013-08-01      31.511
         Lahore          2013-09-01         NaN

[330 rows x 2 columns]


Multi-level indexes can make it easy to comprehend your dataset when one category is nested inside another category.  
  
Some examples on sorting by index values.

In [30]:
# Sort temperatures_ind by index values
print(temperatures_ind.sort_index())

# Sort temperatures_ind by index values at the city level
print(temperatures_ind.sort_index(level="city"))

# Sort temperatures_ind by country then descending city
print(temperatures_ind.sort_index(level=["country","city"],ascending=[True,False]))

                          date  avg_temp_c
country     city                          
Afghanistan Kabul   2000-01-01       3.326
            Kabul   2000-02-01       3.454
            Kabul   2000-03-01       9.612
            Kabul   2000-04-01      17.925
            Kabul   2000-05-01      24.658
...                        ...         ...
Zimbabwe    Harare  2013-05-01      18.298
            Harare  2013-06-01      17.020
            Harare  2013-07-01      16.299
            Harare  2013-08-01      19.232
            Harare  2013-09-01         NaN

[16500 rows x 2 columns]
                             date  avg_temp_c
country       city                           
Côte D'Ivoire Abidjan  2000-01-01      27.293
              Abidjan  2000-02-01      27.685
              Abidjan  2000-03-01      29.061
              Abidjan  2000-04-01      28.162
              Abidjan  2000-05-01      27.547
...                           ...         ...
China         Xian     2013-05-01      18.979
 

Sorting index values is similar to sorting values in columns, except that you call `.sort_index()` instead of `.sort_values()`.  
  
**Slicing and subsetting with .loc and .iloc**  
Slicing is a technique to select consecutive elements from objects.  
  
**slicing lists**  
By providing first and last position, separated by a colon `:` in squared brackets`[]`.
  
  `df[2:5]`  
  `df[:5]` returns everything from start to last position  
  `df[:]` returns the whole list

In [3]:
beers = ["lager","ale","pilsner","stout","porter","barley wine","sahti"]

# to slice the list
print(beers[2:5])
print(beers[:5])
print(beers[:])

['pilsner', 'stout', 'porter']
['lager', 'ale', 'pilsner', 'stout', 'porter']
['lager', 'ale', 'pilsner', 'stout', 'porter', 'barley wine', 'sahti']


Sorting dataframes  
  
**Sort the index before we slice**  

In [6]:
import pandas as pd
# OMA need to read in the data first
path3 = "data/temperatures.csv"
temperatures = pd.read_csv(path3, index_col=0)
print(temperatures.shape)
temperatures.head()

(16500, 4)


date     city        country  avg_temp_c
0  2000-01-01  Abidjan  Côte D'Ivoire      27.293
1  2000-02-01  Abidjan  Côte D'Ivoire      27.685
2  2000-03-01  Abidjan  Côte D'Ivoire      29.061
3  2000-04-01  Abidjan  Côte D'Ivoire      28.162
4  2000-05-01  Abidjan  Côte D'Ivoire      27.547

In [11]:
# sort index
temp_srt = temperatures.set_index(["date","city"]).sort_index()
temp_srt.head()

country  avg_temp_c
date       city                                  
2000-01-01 Abidjan      Côte D'Ivoire      27.293
           Addis Abeba       Ethiopia      17.391
           Ahmadabad            India      20.781
           Aleppo               Syria       5.205
           Alexandria           Egypt      13.579

In above example the index is a multilevel index of `date` and `city` then the index is sorted using `sort_index()` method. To slice outer level of the index we can call `.loc["value_1:"value_2"]` passing the first and last values separated by colon.

In [12]:
temp_srt.loc["2000-01-01":"2000-02-01"]

country  avg_temp_c
date       city                                  
2000-01-01 Abidjan      Côte D'Ivoire      27.293
           Addis Abeba       Ethiopia      17.391
           Ahmadabad            India      20.781
           Aleppo               Syria       5.205
           Alexandria           Egypt      13.579
...                               ...         ...
2000-02-01 Tokyo                Japan       1.765
           Toronto             Canada      -4.177
           Umm Durman           Sudan      24.991
           Wuhan                China       5.842
           Xian                 China       0.943

[200 rows x 2 columns]

The differences compare to slicings a `list` is:  
1. Rather than slicing row numbers we specify index values  
2. The final value is included.  
  
The same technique does not work on the inner index level.  
  
**Slicing the inner index levels badly**  
Below example gives an empty dataframe and pandas does not throw an error.  
  
The correct way is to  
**Slicing the inner index levels correctly**  

In [15]:
# Bad example - this returns an empty dataframe instead of the three cities we wanted
temp_srt.loc["Abidjan":"Ahmadabad"]  

Empty DataFrame
Columns: [country, avg_temp_c]
Index: []

In [17]:
# Good example - 
temp_srt.loc[("2000-01-01","Abidjan"):("2001-01-01","Ahmadabad")]

country  avg_temp_c
date       city                                  
2000-01-01 Abidjan      Côte D'Ivoire      27.293
           Addis Abeba       Ethiopia      17.391
           Ahmadabad            India      20.781
           Aleppo               Syria       5.205
           Alexandria           Egypt      13.579
...                               ...         ...
2000-12-01 Wuhan                China       7.588
           Xian                 China       1.265
2001-01-01 Abidjan      Côte D'Ivoire      26.920
           Addis Abeba       Ethiopia      17.454
           Ahmadabad            India      19.770

[1203 rows x 2 columns]

Above we are passing a `tuple` as first argument to `loc[]`.  
  
Slicing columns  
`df.loc[:, "col_1":"col_2"]`  
Here a colon `:` by itself means keep everything, for the rows in this case. The second name takes column_names as first and last arguments to slice on.  
  
We can slice based both on rows and columns at the same time by just passing the slices to each argument.

In [19]:
# slice all rows and specific columns
temp_srt.loc[:, "country":"avg_temp_c"].head()

country  avg_temp_c
date       city                                  
2000-01-01 Abidjan      Côte D'Ivoire      27.293
           Addis Abeba       Ethiopia      17.391
           Ahmadabad            India      20.781
           Aleppo               Syria       5.205
           Alexandria           Egypt      13.579

In [25]:
# Slice twice
temp_srt.loc[("2000-01-01","Aleppo"):("2000-01-02","Berlin"),"country":"avg_temp_c"]

country  avg_temp_c
date       city                          
2000-01-01 Aleppo       Syria       5.205
           Alexandria   Egypt      13.579
           Ankara      Turkey      -4.621
           Baghdad       Iraq      10.024
           Bangalore    India      23.673
...                       ...         ...
           Tokyo        Japan       3.660
           Toronto     Canada      -7.373
           Umm Durman   Sudan      23.375
           Wuhan        China       3.329
           Xian         China      -2.819

[97 rows x 2 columns]

Above we have previous two slices being performed in one line of code.  
  
**Slicing by dates**  
We can do this as we have seen by using the dates.  
`df.loc["yyyy-mm-dd":"yyyy-mm-dd"]`  
or by using partial dates like `yyyy`:`yyyy` with no month or day parts  
`df.loc["yyyy":"yyyy"]`

In [28]:
# using dates
temp_srt.loc["2000-01-01":"2001-12-31"]

country  avg_temp_c
date       city                                  
2000-01-01 Abidjan      Côte D'Ivoire      27.293
           Addis Abeba       Ethiopia      17.391
           Ahmadabad            India      20.781
           Aleppo               Syria       5.205
           Alexandria           Egypt      13.579
...                               ...         ...
2001-12-01 Tokyo                Japan       3.731
           Toronto             Canada      -0.073
           Umm Durman           Sudan      25.243
           Wuhan                China       4.872
           Xian                 China      -1.691

[2400 rows x 2 columns]

In [27]:
# using partial dates
temp_srt.loc["2000":"2001"]

country  avg_temp_c
date       city                                  
2000-01-01 Abidjan      Côte D'Ivoire      27.293
           Addis Abeba       Ethiopia      17.391
           Ahmadabad            India      20.781
           Aleppo               Syria       5.205
           Alexandria           Egypt      13.579
...                               ...         ...
2000-12-01 Tokyo                Japan       4.245
           Toronto             Canada      -8.854
           Umm Durman           Sudan      23.539
           Wuhan                China       7.588
           Xian                 China       1.265

[1200 rows x 2 columns]

Pandas interprets this as all dates in `2000` and `2001`.  
  
**Subsetting by row/column number**  
We can also subsetting a dataframe by using the `iloc` method.  
`df.iloc[2:5, 1:4]`  
This is similar syntax as for subsetting lists, except that there are two arguments, one for `rows` and one for `columns`. Note that like list-slicing, unlike `loc`, the final values are not included in the slice, in this case row `5` and column `4` are not included.  
  
Some more examples:

In [ ]:
# OMA, set index
temperatures_ind = temperatures.set_index(["country","city"])

# Sort the index of temperatures_ind
temperatures_srt = temperatures_ind.sort_index()

# Subset rows from Pakistan to Russia
print(temperatures_srt.loc["Pakistan":"Russia"])

# Try to subset rows from Lahore to Moscow
print(temperatures_srt.loc["Lahore":"Moscow"])

# Subset rows from Pakistan, Lahore to Russia, Moscow
print(temperatures_srt.loc[("Pakistan","Lahore"):("Russia","Moscow")])

Combining slicing with .loc[] provides a concise syntax for subsetting.  
  
More examples:

In [45]:
# Subset rows from India, Hyderabad to Iraq, Baghdad
print(temperatures_srt.loc[("India","Hyderabad"):("Iraq","Baghdad")])

# Subset columns from date to avg_temp_c
print(temperatures_srt.loc[:,"date":"avg_temp_c"])

# Subset in both directions at once
print(temperatures_srt.loc[("India","Hyderabad"):("Iraq","Baghdad"),"date":"avg_temp_c"])

                         date  avg_temp_c
country city                             
India   Hyderabad  2000-01-01      23.779
        Hyderabad  2000-02-01      25.826
        Hyderabad  2000-03-01      28.821
        Hyderabad  2000-04-01      32.698
        Hyderabad  2000-05-01      32.438
...                       ...         ...
Iraq    Baghdad    2013-05-01      28.673
        Baghdad    2013-06-01      33.803
        Baghdad    2013-07-01      36.392
        Baghdad    2013-08-01      35.463
        Baghdad    2013-09-01         NaN

[2145 rows x 2 columns]
                          date  avg_temp_c
country     city                          
Afghanistan Kabul   2000-01-01       3.326
            Kabul   2000-02-01       3.454
            Kabul   2000-03-01       9.612
            Kabul   2000-04-01      17.925
            Kabul   2000-05-01      24.658
...                        ...         ...
Zimbabwe    Harare  2013-05-01      18.298
            Harare  2013-06-01      17.020

Slicing with `.loc[]` lets you take subsets in both directions at once.  
  
More examples:

In [46]:
# Use Boolean conditions to subset temperatures for rows in 2010 and 2011
temperatures_bool = temperatures[(temperatures["date"] >= "2010-01-01") & (temperatures["date"] <= "2011-12-31")]
print(temperatures_bool)

# Set date as the index and sort the index
temperatures_ind = temperatures.set_index("date").sort_index()

# Use .loc[] to subset temperatures_ind for rows in 2010 and 2011
print(temperatures_ind.loc["2010":"2011"])

# Use .loc[] to subset temperatures_ind for rows from Aug 2010 to Feb 2011
print(temperatures_ind.loc["2010-08":"2011-02"])

             date     city        country  avg_temp_c
120    2010-01-01  Abidjan  Côte D'Ivoire      28.270
121    2010-02-01  Abidjan  Côte D'Ivoire      29.262
122    2010-03-01  Abidjan  Côte D'Ivoire      29.596
123    2010-04-01  Abidjan  Côte D'Ivoire      29.068
124    2010-05-01  Abidjan  Côte D'Ivoire      28.258
...           ...      ...            ...         ...
16474  2011-08-01     Xian          China      23.069
16475  2011-09-01     Xian          China      16.775
16476  2011-10-01     Xian          China      12.587
16477  2011-11-01     Xian          China       7.543
16478  2011-12-01     Xian          China      -0.490

[2400 rows x 4 columns]
                  city    country  avg_temp_c
date                                         
2010-01-01  Faisalabad   Pakistan      11.810
2010-01-01   Melbourne  Australia      20.016
2010-01-01   Chongqing      China       7.921
2010-01-01   São Paulo     Brazil      23.738
2010-01-01   Guangzhou      China      14.136
...  

Using `.loc[]` in conjunction with a date index provides an easy way to subset for rows before or after some date.  
  
Some examples using `iloc[]`

In [47]:
# Get 23rd row, 2nd column (index 22, 1)
print(temperatures.iloc[22:23,1:2])

# Use slicing to get the first 5 rows
print(temperatures.iloc[0:5,])

# Use slicing to get columns 3 to 4
print(temperatures.iloc[:,2:5])

# Use slicing in both directions at once
print(temperatures.iloc[0:5,2:5])

       city
22  Abidjan
         date     city        country  avg_temp_c
0  2000-01-01  Abidjan  Côte D'Ivoire      27.293
1  2000-02-01  Abidjan  Côte D'Ivoire      27.685
2  2000-03-01  Abidjan  Côte D'Ivoire      29.061
3  2000-04-01  Abidjan  Côte D'Ivoire      28.162
4  2000-05-01  Abidjan  Côte D'Ivoire      27.547
             country  avg_temp_c
0      Côte D'Ivoire      27.293
1      Côte D'Ivoire      27.685
2      Côte D'Ivoire      29.061
3      Côte D'Ivoire      28.162
4      Côte D'Ivoire      27.547
...              ...         ...
16495          China      18.979
16496          China      23.522
16497          China      25.251
16498          China      24.528
16499          China         NaN

[16500 rows x 2 columns]
         country  avg_temp_c
0  Côte D'Ivoire      27.293
1  Côte D'Ivoire      27.685
2  Côte D'Ivoire      29.061
3  Côte D'Ivoire      28.162
4  Côte D'Ivoire      27.547


Use `.iloc[]` to specify a subset using the row or column numbers.  
  
**Working with Pivot tables**  
TBD

In [ ]:
#https://campus.datacamp.com/courses/data-manipulation-with-pandas/slicing-and-indexing-dataframes?ex=11